cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go

```diff
@@ -37,6 +37,53 @@ const (
 	NodesKubeadmConfigClusterRoleName = "kubeadm:nodes-kubeadm-config"
 )
 
+// UploadRemoveConfiguration uploads a new configuration with the client removed
+func UploadRemoveConfiguration(cfg *kubeadmapi.InitConfiguration, client clientset.Interface) error {
```

Comment:
I agree with you this function should be refactored. What about:

- Read the ConfigMap (prior art [here](https://github.com/kubernetes/kubernetes/blob/c6977e5aac7b2c3645825b4b21512d4b93704cfa/cmd/kubeadm/app/util/config/cluster.go#L236))
- Unmarshal the ClusterStatus bit (reuse [unmarshalClusterStatus](https://github.com/kubernetes/kubernetes/blob/c6977e5aac7b2c3645825b4b21512d4b93704cfa/cmd/kubeadm/app/util/config/cluster.go#L252) func ; requires make it public)
- Delete the node entry from ClusterStatus
- Marshal back the ClusterStatus and Save

The main advantage of this approach is that we are not touching at all the ClusterConfiguration part.
Wdyt

Reply:
Additionally, we should take only a node name string (instead of an InitConfiguration) as a func parameter to achieve what @fabriziopandini suggested.